### DISCUSSION TITLE
How to disable response headers name to be converted to lower case

### DISCUSSION THREAD

**Discussion by Gelembjuk** [fastapi/fastapi#12795] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Response

app = FastAPI()


@app.get("/headers-and-object/")
def get_headers(response: Response):
    response.headers["X-Cat-Dog"] = "alone in the world"
    return {"message": "Hello World"}

# i run this with
fastapi dev test.py --host 0.0.0.0 --port 8080

# the response is
curl -i "http://0.0.0.0:8080/headers-and-object/"
HTTP/1.1 200 OK
date: Fri, 08 Nov 2024 11:43:54 GMT
server: uvicorn
content-length: 25
content-type: application/json
x-cat-dog: alone in the world

{"message":"Hello World"}
```

### Description

The problem is that response headers are converted to lower case (names of headers). 
I need to keep a header without changes. I have no option to live with lower case headers. I have some JS application that reads responses and that app can not be modified and it expects a header is as. It can not find a lower case header.

The header is "x-cat-dog" but i need it to be same as i returned "X-Cat-Dog"

What can i do with this? Are there any options?

### Operating System

Linux

### FastAPI Version

0.115.2

### Pydantic Version

2.9.2

### Python Version

3.11

**Comment by luzzodev** [comment, marked as answer]
Hi @Gelembjuk unfortunally is not possible to do it.

Explaination:

1. According to RFC for HTTP1.x in any case headers key **must** be case insensitive and a common implementation for performance reason is to lowercase it.
2. According to RFC for HTTP2.x headers key  **must** be case insensitive and  **must** be lowercase, so it gives more sense for an implementation to workwith lowercase. In HTTP 2.x not lowercase key should be dropped or refused.
3. Headers, MutableHeader class in yuor case, comes directly from startlette and not from Fastapi.

With that in mind your js application do not follow HTTP RFC.

Just out of pure curiosity, and I discourage its real application, you can try the following approach:

Create your custom MutableHeaders class that inherits from startlette.datastructures.MutableHeaders 
Create your custom Response class that inherits from startlette.response.Response

And in both custom classes implement a version for case insensitive header key without converting them in lowercase.

**Reply by Gelembjuk** [reply]
Ok. Thanks for the response.
If i create that custom classes, how can i inject them to fastapi to use instead of a standard classes? is there any documentation regarding this?

**Reply by luzzodev** [reply]
Check how to create and use [CustomResponses](https://fastapi.tiangolo.com/advanced/custom-response/#custom-response-class)

**Reply by Gelembjuk** [reply]
thanks. i will try to play with this

**Comment by jevou** [comment]
Hello @Gelembjuk, sorry to bother. I am facing the same problem, would you mind sharing your solution? I am for now talking to my backend through nginx reverse proxy.

**Reply by Gelembjuk** [reply]
Hi.
I didn't solve it. 
In my app i seems did some workaround to live with lower case headers.

**Reply by jevou** [reply]
OK, thank you for reply

**Comment by kasimov-maxim** [comment]
try this solution using middleware that modifies headers at the ASGI level before sending:

```python
from starlette.types import ASGIApp, Receive, Scope, Send

# ...

class CapitalizeHeadersMiddleware:

    def __init__(self, app: ASGIApp):
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                headers = []
                for name, value in message["headers"]:
                    # Capitalize headers:
                    #   'x-cat-dog' -> 'x-Cat-Dog'
                    decoded_name = name.decode("latin1")
                    capitalized = "-".join(
                        part.capitalize() for part in decoded_name.split("-")
                    )
                    headers.append((capitalized.encode("latin1"), value))
                message["headers"] = headers
            await send(message)

        await self.app(scope, receive, send_wrapper)

# ...

app = FastAPI()
app.add_middleware(CapitalizeHeadersMiddleware)

```

**Reply by jevou** [reply]
appreciate it, thank you 🙏
